firehose.go

```diff
@@ -0,0 +1,93 @@
+package hedwig
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"io"
+)
+
+type Firehose struct {
+	messageValidator *messageValidator
+}
+
+func (f *Firehose) Deserialize(reader io.Reader) ([]Message, error) {
+	// TODO: loop through contents and return messages
+	var messagePayload []byte
+	runWithTransportMessageAttributes := false
+	var messages []Message
+	if f.messageValidator.encoder.IsBinary() {
+		for {
+			// TLV format: 8 bytes for size of message, n bytes for the actual message
+			msgSize := make([]byte, 8)
+			_, err := reader.Read(msgSize)
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, err
+			}
+			msgLength := binary.LittleEndian.Uint64(msgSize)
+			messagePayload = make([]byte, msgLength)
+			_, err = reader.Read(messagePayload)
```

Comment:
Also, additionally, there's this edge case:
> If some data is available but not len(p) bytes, Read conventionally returns what is available instead of waiting for more.

Which means if you _need_ to read `n` bytes, you may have to call `Read` multiple times until the _total_ bytes read = n. However, I feel like GCS implementation of reader wouldn't run into this so might be okay to handle as a separate diff.

Reply:
ah I see will add a test for this too

Reply:
hmm so I was thinking about this and I think in the case of Binary TLV data if the first read returns EOF but still has some data the file might be malformed since the first read is for length of message so based on the format it would make no sense for returned bytes and EOF in that first read

I think you are correct though in the non binary new line seperated case (line 45) that we have to handle the returned bytes before EOF

Reply:
Yeah for the binary format, the second read returning eof and n>0 would be valid. 

Of course this all becomes moot if scanner is a better api?

Reply:
ah yeah I think that is handled in line 32 since I have the code continue if err == io.EOF

Reply:
ah right. Do you also need a `if err == EOF { break }` after the deserialization ? Or we relying on the `Read` call in the next loop to exit it instead

Reply:
I'm rely on the EOF on the first read but yeah, will have to test to make sure

Reply:
okay updated the diff, I just manually tested this by ensuring that subsequent calls to Read() after EOF still return EOF which it does